### PR TITLE
280/token list rounding

### LINKED ISSUE #280: Wrong balance and rounding
# Description
I encountered some issues with the displayed balances
wrongly displayed volumes
rounding issues

# Procedure
 on RINKEBY
- I connected my MM account with Swap
- looked at the token list in the **from** section
- checked balances

# Outcome

1. Wrong balance

The Token displayed here is [USDT](https://rinkeby.etherscan.io/token/0xd9ba894e0097f8cc2bbc9d24d308b98e36dc6d02?a=0x52df85e9de71aa1c210873bcf37ec46d36c99dc2)

MM balance
Bildschirmfoto 2021-03-18 um 12 31 31

GP SWAP balance

Bildschirmfoto 2021-03-18 um 12 31 43
Bildschirmfoto 2021-03-18 um 12 31 59
I can trade the max volume
Bildschirmfoto 2021-03-18 um 12 32 06

Note: I got the tokens during this [trade](https://pr307--gpui.review.gnosisdev.com/rinkeby/orders/0x2617DBB9995548C6E10203DA2E17D6563D4C9DCAB63E4A8EBF62E3482FE4092B52DF85E9DE71AA1C210873BCF37EC46D36C99DC26040AC52)

2. Rounding
If i noticed it correctly the balance is always shown with 4 significant digits

Bildschirmfoto 2021-03-18 um 12 38 20
Bildschirmfoto 2021-03-18 um 12 38 27

Shouldn't it be 248 [DAI](https://rinkeby.etherscan.io/address/0x5592ec0cfb4dbc12d3ab100b257153436a1f0fea) or 247.96 DAI if it is rounded? 

Also for higher volumes 
MM account:
Bildschirmfoto 2021-03-18 um 12 19 28
Swap Balance:
Bildschirmfoto 2021-03-18 um 12 21 38

Shouldn't it be 157800....[DAI](https://rinkeby.etherscan.io/address/0xc7ad46e0b8a400bb3c915120d284aafba8fc4735)

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Closes #280 

Showing lots more precision on token list dropdown balances:
screenshot_2021-04-05_10-47-12

~Small change surgically inserting a `const` into original code. Trying out a different approach than the `mod` files.~
Using the `Mod` files approach

# Testing
1. Having balance in at least one token, open the token selector dropdown
2. Wait for all balances to load
3. Verify your tokens are displayed with the highest precision available

**Note:** 
- At most the 7 most relevant digits will be displayed.
- If there are more than 7, a `...` will be appended to the end
- That applies to both values below and above `0`.

### CODE REVIEW

**Comment by alfetopito** [comment]
Regarding the failed build, there's currently an issue DevOps is aware of with GH actions.

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+import { LightGreyCard } from 'components/Card'
+import TokenListLogo from 'assets/svg/tokenlist.svg'
+import QuestionHelper from 'components/QuestionHelper'
+import useTheme from 'hooks/useTheme'
+
+function currencyKey(currency: Currency): string {
+  return currency instanceof Token ? currency.address : currency === ETHER ? 'ETHER' : ''
+}
+
+export const StyledBalanceText = styled(Text)`
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:29]: Mod file: added `export`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  otherSelected,
+  style,
+  BalanceComponent = Balance // gp-swap added
+}: {
+  currency: Currency
+  onSelect: () => void
+  isSelected: boolean
+  otherSelected: boolean
+  style: CSSProperties
+  BalanceComponent?: (params: { balance: CurrencyAmount }) => JSX.Element // gp-swap added
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:113]: Mod file: new parameter `BalanceComponent`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+          {currency.symbol}
+        </Text>
+        <TYPE.darkGray ml="0px" fontSize={'12px'} fontWeight={300}>
+          {currency.name} {!isOnSelectedList && customAdded && '• Added by user'}
+        </TYPE.darkGray>
+      </Column>
+      <TokenTags currency={currency} />
+      <RowFixed style={{ justifySelf: 'flex-end' }}>
+        {/* {balance ? <Balance balance={balance} /> : account ? <Loader /> : null} */}
+        {balance ? <BalanceComponent balance={balance} /> : account ? <Loader /> : null}
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:143]: Mod file: using new parameter `BalanceComponent`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+  currencies: Currency[]
+  selectedCurrency?: Currency | null
+  onCurrencySelect: (currency: Currency) => void
+  otherCurrency?: Currency | null
+  fixedListRef?: MutableRefObject<FixedSizeList | undefined>
+  showETH: boolean
+  showImportView: () => void
+  setImportToken: (token: Token) => void
+  breakIndex: number | undefined
+  BalanceComponent?: (params: { balance: CurrencyAmount }) => JSX.Element // gp-swap added
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:172]: Mod file: new parameter `BalanceComponent`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        )
+      } else {
+        return (
+          <CurrencyRow
+            style={style}
+            currency={currency}
+            isSelected={isSelected}
+            onSelect={handleSelect}
+            otherSelected={otherSelected}
+            BalanceComponent={BalanceComponent} // gp-swap added
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:236]: Mod file: passing `BalanceComponent` to `CurrencyRow`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+      chainId,
+      inactiveTokens,
+      onCurrencySelect,
+      otherCurrency,
+      selectedCurrency,
+      setImportToken,
+      showImportView,
+      breakIndex,
+      theme.text1,
+      BalanceComponent // gp-swap added
```
Comment (alfetopito) [inline src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:251]: Mod file: useCallback dependency on `BalanceComponent`

File: src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx (hunk 3 above)
Comment (W3stside) [reply src/custom/components/SearchModal/CurrencyList/CurrencyListMod.tsx:143]: You could also keep the same name and only change the mod by adding the new prop

File: src/custom/components/SearchModal/CurrencyList/index.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import { CurrencyAmount } from '@uniswap/sdk'
+
+import { LONG_PRECISION } from 'custom/constants'
```
Comment (W3stside) [inline src/custom/components/SearchModal/CurrencyList/index.tsx:4]: can just be `constants/index` - i believe that's the way we have it standardised - avoiding adding `custom` in the path

File: src/custom/components/SearchModal/CurrencyList/index.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import CurrencyListMod, { StyledBalanceText } from './CurrencyListMod'
+
+export function Balance({ balance }: { balance: CurrencyAmount }) {
+  return <StyledBalanceText title={balance.toExact()}>{balance.toSignificant(LONG_PRECISION)}</StyledBalanceText>
+}
+
+export default function CurrencyList(
+  ...params: Parameters<typeof CurrencyListMod>
+): ReturnType<typeof CurrencyListMod> {
+  return <CurrencyListMod {...params[0]} BalanceComponent={Balance} />
```
Comment (W3stside) [inline src/custom/components/SearchModal/CurrencyList/index.tsx:15]: hard to read :see_no_evil: but fine if it works idgaf :D

**Review by W3stside — APPROVED** [review]
so much work for such a simple change, i hate modding!! nice work. and thanks for changing it. maybe in the future we can indeed make simple changes like this directly but let's just play it safe for now :zap:

**Review by biocom — APPROVED** [review]

File: src/custom/components/SearchModal/CurrencyList/index.tsx (hunk 8 above)
Comment (alfetopito) [reply src/custom/components/SearchModal/CurrencyList/index.tsx:15]: Any suggestion?

File: src/custom/components/SearchModal/CurrencyList/index.tsx (hunk 7 above)
Comment (alfetopito) [reply src/custom/components/SearchModal/CurrencyList/index.tsx:4]: https://github.com/gnosis/gp-swap-ui/pull/386

File: src/custom/components/SearchModal/CurrencyList/index.tsx (hunk 8 above)
Comment (alfetopito) [reply src/custom/components/SearchModal/CurrencyList/index.tsx:15]: https://github.com/gnosis/gp-swap-ui/pull/386

**Pull request merged**